design/validation-spi.md

```diff
@@ -0,0 +1,293 @@
+    // ...
+    String string = (String) input;
+    if (string.trim().length() == 0) {
+        context.addError(new ValidationError(ID, inputHint, MESSAGE_BLANK, input));
+    }
+    return context;
+}
+```
+### Rendering Error Messages
+
+Validation errors should provide a internationalizable message (key) for `ResourceBundle` lookups as well as optional 
+message parameters to provide information about the validated input as well as the valdiated value.
+
+The proposed `ValidationError` class provides a `formatMessage` method which takes a `BiFunction` to abstract
+the actual error message rendering. The first parameter is the `message` key followed by an array of message parameters.
+
+The first message parameter will be the `inputHint` passed into the `Validator#validate(..)` followed by the `messageParameters`.
+
+```java
+public String formatMessage(BiFunction<String, Object[], String> formatter) {
+    Objects.requireNonNull(formatter, "formatter must not be null");
+    return formatter.apply(message, getInputHintWithMessageParameters());
+}
+```
+
+This allows error message rendering like:
+```
+error-invalid-value={0} is invalid: <{1}>  --> address.zip is invalid: <null>
+```
+
+Here is an [example of a custom error message rendering](https://github.com/keycloak/keycloak/pull/7887/files#diff-e8a31e5e873f1c5826f3c2ce581056a85457d21bc60504d06873410ef36f9bcbR119) with the proposed API, look for the `public void formatError()` test.
+
+### Suggest Built-in Validations
+
+It is suggested to provide a set of built-in validations.
+
+1. Length - Checks if a string certain length
+1. NotEmpty - Checks if a string, collection, map is not empty
+1. Number - Checks if a given value is a number (Integer / Double)
+1. Pattern - Checks if the given value matches a pattern
+1. URI - Checks if the given value is a valid URI
+1. Email - Checks if the given value is a valid email address
+
+### Integration plan
+
+1. Prepare Validation SPI in `server-spi-private` in the `org.keycloak.validation` package and consolidate the other Validation APIs to use the new Validation SPI. Note, that most of the validation code can be directly replaced.
+1. Adapt User-Profile support to use the new Validation SPI
+1. Promote Validation SPI to `server-spi` as an official API and expose validator lookup via KeycloakSession
+
+### Open Questions
+1. Should users be able to override built-in Valdators?
+1. How should the error message keys look like?
+1. Do we need a concept like a `validation hook` to allow running user provided Validators before/after/instead of built-in validators?
+1. Should ValidationErrors be Comparable? Do we need concepts like `order` or `priority`?
+
+### Further Work
+1. Add support to report `ValidationWarning`s next to `ValidationError`s to report problems, that don't fail the validation. A generic base type could be `ValidationProblem`.
```

Comment:
It doesn't need to be the validators responsibility to decide if sth. is a warning or an error. Same validator could be called with different configurations => We only need one ValidationError but one validatar config for warnings and one for errors.
Example: A phone number could be validated with min length 3 max mlength 20 to produce warning on a warning on length < 3. This is just a hint that some totally wrong value might be put into the field.
A second validation uses just max max length = 100. Because this is a strict limit.

Reply:
>It doesn't need to be the validators responsibility to decide if sth. is a warning or an error.

+1

What would be the purpose of warning anyway? What's the use case here?